File: Yandex/Practicum/Algorightms_and_Data_Structures/6_sprint/J._Топологическая_сортировка/solution.py
import sys


def dfs(source: int, adj_list: dict[int, list[int]], colors: list[str], output: list[int]) -> None:
    call_stack: list[int] = [source]
    while call_stack:
        vertex: int = call_stack.pop()
        if colors[vertex] == 'white':
            colors[vertex] = 'grey'
            call_stack.append(vertex)
            for neighbor in adj_list.get(vertex, []):
                if colors[neighbor] == 'white': call_stack.append(neighbor)
        elif colors[vertex] == 'grey':
            output.append(vertex)
            colors[vertex] = 'black'


def solution() -> None:
    """
    Time Complexity O(V + E)
    Memory Complexity O(V + E)
    """
    n, m = map(int, sys.stdin.readline().rstrip().split())
    adj_list: dict[int, list[int]] = dict()
    for _ in range(m):
        before, after = map(int, sys.stdin.readline().rstrip().split())
        if before not in adj_list: adj_list[before] = []
        adj_list[before].append(after)
    colors: list[str] = ['white'] * (n + 1)
    output: list[int] = []
    for vertex in range(n, 0, -1):
        if colors[vertex] == 'white': dfs(vertex, adj_list, colors, output)
    output.reverse()
    sys.stdout.write(' '.join(str(vertex) for vertex in output) + '\n')


if __name__ == '__main__':
    solution()

File: Yandex/Practicum/Algorightms_and_Data_Structures/6_sprint/J._Топологическая_сортировка/solution.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <string>


void dfs(int& source, std::unordered_map<int, std::vector<int>>& adjList, std::vector<std::string>& colors, std::vector<int>& output) {
    std::vector<int> callStack = {source};
    while (!callStack.empty()) {
        int vertex = callStack[callStack.size() - 1];
        callStack.pop_back();
        if (colors[vertex] == "white") {
            colors[vertex] = "grey";
            callStack.push_back(vertex);
            for (int& neighbor : adjList[vertex]) {
                if (colors[neighbor] == "white") callStack.push_back(neighbor);
            }
        } else if (colors[vertex] == "grey") {
            output.push_back(vertex);
            colors[vertex] = "black";
        }
    }
}



void solution() {
    /*
    Time Complexity O(V + E)
    Memory Complexity O(V + E)
    */
    int n, m;
    std::cin >> n >> m;
    std::unordered_map<int, std::vector<int>> adjList;
    for (int i = 0; i < m; ++i) {
        int before, after;
        std::cin >> before >> after;
        adjList[before].push_back(after);
    }
    std::vector<std::string> colors (n + 1, "white");
    std::vector<int> output;
    for (int vertex = n; vertex > 0; --vertex) {
        if (colors[vertex] == "white") dfs(vertex, adjList, colors, output);
    }
    for (int idx = n - 1; idx >= 0; --idx) {
        if (idx != n - 1) std::cout << " ";
        std::cout << output[idx];
    }
    std::cout << "\n";
}


int main() {
    solution();
}